### PR TITLE
Backslash record type expression entry key

### PR DESCRIPTION
**Pull request by brettz9**
- fix: disallow backslash in `RecordTypeExprEntryKey` when not paired to be followed by a non-quote character
- fix: tighten unquoted `RecordTypeExprEntryKey` to disallow numbers followed by non-numbers (except as part of a decimal and/or scientific notation)
- fix: allow optional `+` at beginning of decimal literals and optional `e` scientific notation at end (at least TypeScript allows in `RecordTypeExprEntryKey` and presumably elsewhere decimal literals are allowed)
- feat: Preserve quote style for `RecordTypeExprEntryKey`

Note that the change to `DecimalNumberLiteralExpr` impacts `NumberLiteralExpr` which impacts `ValueExpr` which impacts many types up to the top: (`UnionTypeExprOperand`, `PrefixUnaryUnionTypeExprOperand`, `KeyQueryExprOperand`, `SuffixUnaryUnionTypeExprOperand`, `GenericTypeExprOperand`, `GenericTypeExprTypeParamOperand`, `ArrayTypeExprOperand`, `FunctionTypeExprParamOperand`, `FunctionTypeExprReturnableOperand`, and `RecordTypeExprEntryOperand`, `TupleTypeExprOperand`, `ParenthesizedExprOperand`, `VariadicTypeExprOperand`, and `TopTypeExpr`, `TopLevel`). However, I would consider this more of a fix as such notation should really have been allowed where decimals are allowed, at least if TypeScript is any indication.

### CODE REVIEW

File: peg_src/jsdoctype.pegjs [hunk 1]
```diff
@@ -399,7 +399,10 @@ NumberLiteralExpr = value:(BinNumberLiteralExpr / OctNumberLiteralExpr / HexNumb
                   }
 
 
-DecimalNumberLiteralExpr = $("-"? [0-9]+ ("." [0-9]+)?)
+DecimalNumberLiteralExpr = $(("+" / "-")? UnsignedDecimalNumberLiteralExpr)
+
+UnsignedDecimalNumberLiteralExpr = $([0-9]+ ("." [0-9]+)? ("e" ("+" / "-")? [0-9]+)?)
```
Comment (ExE-Boss) [inline peg_src/jsdoctype.pegjs]: This should be its own commit.

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: `UnsignedDecimalNumberLiteralExpr` is needed elsewhere in the commit related to `RecordTypeExprEntryKey` fixes.

I understand not crowding too many things into a PR (like my previously having added the testing refactoring into that other PR), but development will be needlessly slowed down, imo, if we give too much attention to splitting up PRs which are already not too large in content.

I find there is no exact definition for the idea of adding "one thing only" anyways, even if it is a fair guiding principle.

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: If you still want this as a separate PR, I can do so, as I do understand the decimal change spills over into other types (and I've updated the original post to draw more attention to this fact). But it was required in connection with this work.

Comment (ExE-Boss) [reply peg_src/jsdoctype.pegjs]: Fine by me.

Comment (sandersn) [reply peg_src/jsdoctype.pegjs]: This improves the parsing of numbers, right? Can you add a few tests with the numbers that are now parseable?

Comment (brettz9) [reply peg_src/jsdoctype.pegjs]: Good catch. I've added tests. I also liberalized a little further in allowing a decimal without a leading digit and added a test for that.
